### ISSUE TITLE
Include tag reader doesn't refer to default tag readers

### ISSUE THREAD

**Issue by scottlowe** [duct-framework/core#16]
I've been trying out the `0.7.0-alpha6` release of Duct and I've encountered an issue with `merge-default-readers` and the construction of the `#duct/include` tag reader:

https://github.com/duct-framework/core/blob/49533e5416285dd08b5f15f451f18473facf9b98/src/duct/core.clj#L102-L110

The problem is that when the `'duct/include` reader is made in the call to `make-include`, it is only passed the custom readers and doesn't receive the merged-in default Duct tag readers e.g. `#duct/env`.

I discovered this issue when trying the alpha version of Duct against my existing Duct codebase. A top level `.edn` config file included another `.edn` file which also contained a `#duct/env` reader tag. I got an error parsing the included file because a definition of the `#duct/env` tag did not exist.

A simple (naive?) solution would be to merge the default and custom readers up-front before passing them to the `make-include` function:

```clojure
(defn- merge-default-readers [custom-readers]
  (let [readers (merge
                  {'duct/env env/env
                   'duct/resource resource}
                  custom-readers)]
    (assoc readers 'duct/include (make-include readers))))
```

That solves my first issue with the missing `#duct/env` reader for my included config file and it's certainly an improvement. However, I then discovered another issue which is that the `duct/include` reader doesn't have access to it's own definition of the `duct/include` reader.

It looks like the currently supported nested includes would not be supported in new versions of Duct. Now I can appreciate why this might be desirable (by design) and why you might consider nested includes to be adding unwanted complexity to client projects, as well as Duct itself. However, our codebase _does_ currently contain a significant number of nested includes.

The reason we are using nested includes with the existing stable version of Duct is that our code is for a large multi-tenanted application, which also has many configurable feature components that are enabled or disabled for each customer, some of which are custom-configured on a per customer basis. I suspect that it might be one of the larger Duct codebases currently in existence. We believe that nested includes do indeed help to manage configuration complexity in our case.

I don't know if you intend to guarantee nested includes use for future versions of Duct?

If only top level includes will be guaranteed in the future, then we'll need to significantly refactor our code. It would be good to know ahead of time so that we could look at making plans for that.

Playing with some spike code in the REPL, I managed to get nested `#duct/include` tag readers working. I did this by defining the merged readers for an initial pass with a symbol placeholder for the the include function var and then I subsequently redefined the include function var to include a reference to itself, but it's an ugly impure hack. That approach is clearly more complicated and the code may be undesirable to to you. Perhaps reader definition could be removed to a separate step that is run only once before read-config is called?

Alternatively, there are a number of other approaches which could be taken in order to facilitate self-referencing readers for nested includes, using perhaps a top-level readers atom or even a macro.

I'd be happy to make a PR that attempts to tackle these issues but would like to know what the desired direction and approach would be first, if any.

**Comment by weavejester** [comment]
Thanks for the report. This is a bug and it does need to be fixed. I'll see if I can do that tonight.

**Comment by weavejester** [comment]
Fixed in 0.7.0-alpha7

**Comment by scottlowe** [comment]
Wow! So swift! Thanks for the fix 🙇
